applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,576 @@
+
+        # project and damp constraint gradients
+        for constraint in self.constraints:
+            con_id = constraint["identifier"].GetString()
+            conGradientDict = self.communicator.getStandardizedGradient(con_id)
+            gradient_variable = self.constraint_gradient_variables[con_id]["gradient"]
+            WriteDictionaryDataOnNodalVariable(conGradientDict, self.optimization_model_part, gradient_variable)
+
+            if constraint["project_gradient_on_surface_normals"].GetBool():
+                self.model_part_controller.ProjectNodalVariableOnUnitSurfaceNormals(gradient_variable)
+
+            self.model_part_controller.DampNodalSensitivityVariableIfSpecified(gradient_variable)
+
+    # --------------------------------------------------------------------------
+    def __computeBufferValue(self):
+    # compute new buffer size and buffer values
+    	for constraint in self.constraints:
+            identifier = constraint["identifier"].GetString()
+            constraint_value = self.communicator.getStandardizedValue(identifier)
+
+            self.constraint_buffer_variables[identifier]["buffer_value-1"] = self.constraint_buffer_variables[identifier]["buffer_value"]
+
+            if self.optimization_iteration > 1:
+
+                if abs(constraint_value - self.constraint_buffer_variables[identifier]["constraint_value-1"]) > self.constraint_buffer_variables[identifier]["max_constraint_change"]:
+                    self.constraint_buffer_variables[identifier]["max_constraint_change"] = abs(constraint_value - self.constraint_buffer_variables[identifier]["constraint_value-1"])
+
+                self.constraint_buffer_variables[identifier]["buffer_size"] = max(self.constraint_buffer_variables[identifier]["buffer_size_factor"] * self.constraint_buffer_variables[identifier]["max_constraint_change"], 1e-12)
+            self.constraint_buffer_variables[identifier]["lower_buffer_value"] = self.constraint_buffer_variables[identifier]["central_buffer_value"] - self.constraint_buffer_variables[identifier]["buffer_size"]
+            self.constraint_buffer_variables[identifier]["upper_buffer_value"] = self.constraint_buffer_variables[identifier]["central_buffer_value"] + self.constraint_buffer_variables[identifier]["buffer_size"]
+
+            if self.__isConstraintActive(constraint):
+                if constraint["type"].GetString() == "=":
+                    self.constraint_buffer_variables[identifier]["buffer_value"] = min(1 - abs(constraint_value) / self.constraint_buffer_variables[identifier]["buffer_size"], 2.0)
+                else:
+                    self.constraint_buffer_variables[identifier]["buffer_value"] = min( (constraint_value - self.constraint_buffer_variables[identifier]["lower_buffer_value"]) / self.constraint_buffer_variables[identifier]["buffer_size"], 2.0 )
+            else:
+                self.constraint_buffer_variables[identifier]["buffer_value"] = 0.0
```

Comment:
Minor and up to you: this code would become much more readable if you first get the variables from the dict and assign them at the end:

```
_central_buffer_value = self.constraint_buffer_variables[identifier]["central_buffer_value"]
_lower_buffer_value = ...

self.constraint_buffer_variables[identifier]["lower_buffer_value"] = _lower_buffer_value

...